oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenDecoderFactoryTests.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client.oidc.authentication;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.TestClientRegistrations;
+import org.springframework.security.oauth2.core.OAuth2TokenValidator;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
+
+import java.time.Duration;
+import java.util.function.Function;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Rafael Dominguez
+ * @since 5.2
+ */
+public class OidcIdTokenDecoderFactoryTests {
+
+	private ClientRegistration.Builder registration = TestClientRegistrations.clientRegistration()
+			.scope("openid");
+
+	private ClientRegistration clientRegistration;
+
+	private OidcIdTokenDecoderFactory oidcIdDecoderFactory;
+
+	private JwtDecoder jwtDecoder;
+
+	@Before
+	public void setUp() {
+		oidcIdDecoderFactory = new OidcIdTokenDecoderFactory();
+		clientRegistration = registration.build();
+	}
+
+	@Test
+	public void createWhenWithDefaultDecoderTheSuccess() {
+		jwtDecoder = oidcIdDecoderFactory.createDecoder(clientRegistration);
+		assertThat(jwtDecoder).isInstanceOf(NimbusJwtDecoder.class);
+	}
+
+	@Test
+	public void createWhenWithCustomDecoderTheSuccess() {
+		oidcIdDecoderFactory.setJwtValidatorFactory(customJwtValidatorFactory);
+		jwtDecoder = oidcIdDecoderFactory.createDecoder(clientRegistration);
+
+		assertThat(jwtDecoder).isInstanceOf(NimbusJwtDecoder.class);
+	}
+
+	private Function<ClientRegistration, OAuth2TokenValidator<Jwt>> customJwtValidatorFactory = (c) -> {
+		OidcIdTokenValidator idTokenValidator = new OidcIdTokenValidator(c);
+		if (c.getRegistrationId().equals("registration-id1")) {
+			idTokenValidator.setClockSkew(Duration.ofSeconds(30));
+		} else if (c.getRegistrationId().equals("registration-id2")) {
+			idTokenValidator.setClockSkew(Duration.ofSeconds(70));
+		}
+		return idTokenValidator;
+	};
+}
```

Comment:
I feel that the only tests we need here are as follows:

- `setJwtValidatorFactoryWhenNullThenThrowIllegalArgumentException`
- `createDecoderWhenClientRegistrationNullThenThrowIllegalArgumentException`
- `createDecoderWhenJwkSetUriEmptyThenThrowOAuth2AuthenticationException`
- `createDecoderWhenClientRegistrationValidThenReturnDecoder`
- `createDecoderWhenCustomJwtValidatorFactorySetThenApplied`

Can you please add these tests? Please let me know if you need any clarification.

Reply:
Thank you very much, I've submitted the changes